### DISCUSSION TITLE
Nesting Loops to accomplish two tasks

### DISCUSSION THREAD

**Discussion by iamskyknight** [pybricks/support#1341] [Q&A] [answered]
I use the "Constant Speed" Pybricks script on my train layout to great effect.  However, I'm having trouble adding additional logic where I want the color sensor to ALSO check for a unique HSV value ("Red"), and if found STOP the motor, Wait 10 seconds, then resume the constant speed loop.  I can get the script to successfully recognize the custom Red HSV and stop the motor, but after the wait timeout, the motor resumes at 100% speed, and no longer abides by the constant speed loop.  Help!

```python
from pybricks.pupdevices import DCMotor, ColorDistanceSensor
from pybricks.tools import StopWatch
from pybricks.parameters import Port, Color
from pybricks.tools import wait

# Initialize the motor and the sensor.
motor = DCMotor(Port.A)
sensor = ColorDistanceSensor(Port.B)
check_color_interval_ms = 20

# These are the sensor reflection values in this setup.
# Adapt them to match your ambient light conditions.
LIGHT = 57
DARK = 16

# Threshold values. We add a bit of hysteresis to make
# sure we skip extra changes on the edge of each track.
hysteresis = (LIGHT - DARK) / 4
threshold_up = (LIGHT + DARK) / 2 + hysteresis
threshold_down = (LIGHT + DARK) / 2 - hysteresis

# Set up custom RED Tile color.
Color.RED_TILE = Color(353, 97, 87)

# Initial position state.
on_track = True
position = 0

# Desired drive speed in mm per second.
SPEED = 200

# It's two studs (16 mm) for each position increase.
MM_PER_COUNT = 16

# Start a timer.
watch = StopWatch()

while True:

    # Measure the reflection.
    reflection = sensor.reflection()

    # If the reflection exceeds the threshold, increment position.
    if (reflection > threshold_up and on_track) or (reflection < threshold_down and not on_track):
        on_track = not on_track
        position += 1
        
    # Compute the target position based on the time.
    target_count = watch.time() / 1000 * SPEED / MM_PER_COUNT

    # The duty cycle is the position error times a constant gain.
    duty = 2*(target_count - position)

    # Apply the duty cycle.
    motor.dc(duty)

    #Check for RED Tile then Stop and Wait
    color = sensor.hsv()
    if sensor.hsv() == Color.RED_TILE:
        print(color)
        motor.stop()
        wait(10000)
```

**Comment by dlech** [comment]
`senor.hsv()` is never going to match a color exactly. Instead use `sensor.detectable_colors([Color.RED_TILE, Color.NONE])` before the loop to calibrate the color detection and `if sensor.color() == Color.RED_TILE:`

Full example: https://docs.pybricks.com/en/stable/pupdevices/colorsensor.html#changing-the-detectable-colors

**Reply by iamskyknight** [reply]
Thanks for the feedback.  Color detection and matching is not the issue, though...that part works fine.  I can't get the script to resume `motor.dc(duty)` logic after the wait expires.  What instead happens is the motor ramps up to 100% full throttle and the train flies off the track.

**Reply by dlech** [reply]
Do do you need to reset the stopwatch and position after waiting 10 seconds?

**Reply by laurensvalk** [reply, marked as answer]
> Do do you need to reset the stopwatch and position after waiting 10 seconds?

The following is a variation that does not require resetting the position.

----------------

@iamskyknight, I'm assuming you're building [something like this](https://pybricks.com/projects/sets/city/60198-cargo-train/speed-control/).

I think you'd want to pause the stopwatch as the train stands still, so it doesn't continue building up a difference with the target position.

So you could do:

```python

    if sensor.hsv() == Color.RED_TILE:
        watch.pause()
        print(color)
        motor.stop()
        wait(10000)
        watch.resume()

```

**Reply by iamskyknight** [reply]
@laurensvalk That was it!!  I had to pause the stopwatch.  The code snippet you provided did the trick. Thank you so very much!

**Reply by laurensvalk** [reply]
If you have time, we'd love to see a video of the result! :smile:
